Yosemite/Yosemite/Stores/InboxNotesStore.swift

```diff
@@ -71,7 +100,65 @@ private extension InboxNotesStore {
                                  noteID: Int64,
                                  actionID: Int64,
                                  completion: @escaping (Result<InboxNote, Error>) -> ()) {
-        remote.markInboxNoteAsActioned(for: siteID, noteID: noteID, actionID: actionID, completion: completion)
+        remote.markInboxNoteAsActioned(for: siteID, noteID: noteID, actionID: actionID) { [weak self] result in
+            switch result {
+            case .failure(let error):
+                completion(.failure(error))
+
+            case .success(let inboxNote):
+                self?.upsertStoredInboxNotesInBackground(readOnlyInboxNotes: [inboxNote], siteID: siteID, onCompletion: {
+                    completion(.success(inboxNote))
+                })
+            }
+        }
+    }
+}
+
+// MARK: - Storage: InboxNote
+//
+private extension InboxNotesStore {
+
+    /// Updates or Inserts specified Inbox Notes Entities in a background thread
+    /// `onCompletion` will be called on the main thread.
+    ///
+    func upsertStoredInboxNotesInBackground(readOnlyInboxNotes: [Networking.InboxNote],
+                                            siteID: Int64,
+                                            onCompletion: @escaping () -> Void) {
+        let derivedStorage = sharedDerivedStorage
+        derivedStorage.perform { [weak self] in
+            self?.upsertStoredInboxNotes(readOnlyInboxNotes: readOnlyInboxNotes,
+                                         in: derivedStorage,
+                                         siteID: siteID)
+        }
+
+        storageManager.saveDerivedType(derivedStorage: derivedStorage) {
+            DispatchQueue.main.async(execute: onCompletion)
+        }
+    }
+
+    /// Updates or Inserts the specified Inbox Note entities
+    ///
+    func upsertStoredInboxNotes(readOnlyInboxNotes: [Networking.InboxNote],
+                                in storage: StorageType,
+                                siteID: Int64) {
+        for inboxNote in readOnlyInboxNotes {
+            let storageInboxNote: Storage.InboxNote = {
+                if let storedInboxNote = storage.loadInboxNote(siteID: siteID, id: inboxNote.id) {
+                    return storedInboxNote
+                }
+                return storage.insertNewObject(ofType: Storage.InboxNote.self)
+            }()
```

Comment:
nit: could this be simplified like
```suggestion
            let storageInboxNote = storage.loadInboxNote(siteID: siteID, id: inboxNote.id) ?? storage.insertNewObject(ofType: Storage.InboxNote.self)
```